### PR TITLE
[Sparse] Raise exception when expand is called on sparse tensor

### PR DESCRIPTION
**Pull request by malfet**
It's already not working, but this makes error message a bit more readable. I.e. it turns:
```
% python -c "import torch;x=torch.eye(3).to_sparse().expand(3,3)" 
```

from
```
NotImplementedError: Could not run 'aten::as_strided' with arguments from the 'SparseCPU' backend.
```
to
```
RuntimeError: Expand is unsupported for Sparse tensors.
```

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

### CODE REVIEW

**Comment by malfet** [comment]
@cpuhrsch do you know why why `torch.eye(3).to_sparse_csr().is_sparse()` returns false? 
Also, why `python -c "import torch;print(torch.eye(3).to_sparse().stride())"` prints `(0, 0)`, but the same command for sparse_csr tensor raises runtime exception?

**Comment by pearu** [comment]
> do you know why why `torch.eye(3).to_sparse_csr().is_sparse()` returns false?

@malfet This is because `.is_sparse` attribute is for testing COO sparse tensor, not for CSR sparse tensor:
```python
>>> torch.eye(3).to_sparse_csr().is_sparse
False
>>> torch.eye(3).to_sparse_csr().is_sparse_csr
True
```
(`.is_sparse` is not called `.is_sparse_coo` because of BC)

> Also, why `python -c "import torch;print(torch.eye(3).to_sparse().stride())"` prints `(0, 0)`, but the same command for sparse_csr tensor raises runtime exception?

sparse compressed tensor implementation defines `sym_strides_custom` (that raises the runtime exception) while the sparse coo tensor implementation doesn't. IIRC, there was an edge case where coo `stride()` result was actually used but I couldn't find it quickly (enabling the exception for coo tensor `strides` might give a hint). In any case, sparse tensors don't have strides and one should not call the method on sparse tensors in general.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1107,6 +1107,8 @@ Tensor expand(const Tensor& self, c10::IntArrayRef size, bool /*unused*/) {
            "): the number of sizes provided (", size.size(), ") ",
            "must be greater or equal to the number of dimensions in the tensor (",
            self.dim(), ")");
+  TORCH_CHECK(!self.is_sparse() && !self.is_sparse_csr(),
```
Comment (pearu) [inline aten/src/ATen/native/TensorShape.cpp:1110]: Use:
```suggestion
  TORCH_CHECK(!(self.is_sparse() || at::sparse_csr::is_sparse_compressed(self)), 
```
that will also capture CSC/BSR/BSC layout cases.

**Review by pearu — CHANGES_REQUESTED** [review]

**Comment by amjames** [comment]
> > > do you know why why `torch.eye(3).to_sparse_csr().is_sparse()` returns false?
> > > (`.is_sparse` is not called `.is_sparse_coo` because of BC)
> 
> @pearu then perhaps the documentation should be updated, because according to it: [`Tensor.is_sparse` ](https://pytorch.org/docs/stable/generated/torch.Tensor.is_sparse.html) "is `True` if the Tensor uses sparse storage layout, `False` otherwise."


IMO: We should break that compatibility. We are still operating under the `beta` tag, so we can do that. We really just need to strict about BC with changes that affect other APIs (like gradcheck)

The docs also probably need a review to make sure that any place where "sparse layout" refers specifically to "sparse_coo" is clarified.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @malfet!

**Comment by pearu** [comment]
Yes, the docs need an update.

In fact, I think we just deprecate the usage of `.is_sparse` and `.is_sparse_csr` attributes (notice that there are no `.is_sparse_csc/bsr/bsc` attributes) in favor of using the `.layout` attribute. 

That is, instead of `x.is_sparse`, use `x.layout is torch.sparse_coo`, etc.

For example, we often use `x.layout in {torch.sparse_coo, torch.sparse_csr, torch.sparse_csc, torch.sparse_bsr, torch.sparse_bsc}` to test if `x` has any of the sparse layouts.

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**